File: WebContent/extras/c#-processing/Figure.cs

using System.Collections.Generic;

namespace shape_detect
{
    public class Figure
    {
        public const double MAX_MISSMATCH = 10000.0;

        private List<Stroke> strokes;
        private List<Curve> curves;

        public int Length
        {
            get { return strokes.Count; }
        }

        public Stroke this[int index]
        {
            get { return strokes[index]; }
        }

        public Figure()
        {
            strokes = new List<Stroke>();
            curves = new List<Curve>();
        }

        public void Add(Stroke stroke)
        {
            strokes.Add(stroke);
        }

        public void CurveLastStroke()
        {
            if (strokes.Count > 0)
            {
                Curve curve = new Curve(strokes[strokes.Count - 1]);
                curve.Downsample();
                curve.Segment();
                curves.Add(curve);
            }
        }

        public Curve GetCurve(int index)
        {
            if (index < curves.Count)
                return curves[index];
            else return null;
        }

        public void Clear()
        {
            strokes.Clear();
            curves.Clear();
        }

        public double Match(Figure other)
        {
            if (curves.Count == other.curves.Count)
            {
                double match = 0.0;
                for (int i = 0; i < curves.Count; i++)
                {
                    match += curves[i].Match(other.curves[i]);
                }
                return match / curves.Count;
            }
            else
            {
                return MAX_MISSMATCH;
            }
        }
    }
}

File: src/com/scrbl/logic/Figure.java
package com.scrbl.logic;

public class Figure
{
	public static final double MAX_MISSMATCH = 10000.0;

	private java.util.ArrayList<Stroke> strokes;
	private java.util.ArrayList<Curve> curves;
	
	public int getCurvesLength() {
		return curves.size();
	}

	public final int getLength()
	{
		return strokes.size();
	}

	public final Stroke getItem(int index)
	{
		return strokes.get(index);
	}

	public Figure()
	{
		strokes = new java.util.ArrayList<Stroke>();
		curves = new java.util.ArrayList<Curve>();
	}

	public final void Add(Stroke stroke)
	{
		strokes.add(stroke);
	}

	public final void CurveLastStroke()
	{
		if (strokes.size() > 0)
		{
			System.out.println("STROKES SIZE : "+strokes.size());
			Curve curve = new Curve(strokes.get(strokes.size() - 1));
			curve.Downsample();
			curve.Segment();
			curves.add(curve);
		}
	}

	public final Curve GetCurve(int index)
	{
		if (index < curves.size())
		{
			return curves.get(index);
		}
		else
		{
			return null;
		}
	}

	public final void Clear()
	{
		strokes.clear();
		curves.clear();
	}

	public final double Match(Figure other)
	{
		if (curves.size() == other.curves.size())
		{
			double match = 0.0;
			for (int i = 0; i < curves.size(); i++)
			{
				match += curves.get(i).Match(other.curves.get(i));
			}
			return match / curves.size();
		}
		else
		{
			return MAX_MISSMATCH;
		}
	}
}